src/models/Transaction.ts

```diff
@@ -33,27 +47,39 @@ export class Transaction {
     public extensions: any[] = [];
 
     @Exclude()
-    private readonly chainId: Buffer;
-
-    constructor(blockData: BlockData, ops: BaseOperation[], chainId: string) {
-        this.blockData = blockData;
-        this.operations = ops;
-        this.chainId = Buffer.from(chainId, "hex");
-        this.refBlockNum = this.blockData.refBlockNum;
-        this.refBlockPrefix = this.blockData.refBlockPrefix;
-        this.expiration = this.blockData.expiration;
-    }
+    public chainId: string;
+
+    /**
+     * Generate signature on transaction data. May return undefined if the signature is not valid for DCore.
+     *
+     * @param key private key
+     * @param chainId id of the DCore chain, different for live/testnet/custom net...
+     */
+    public signature(key: ECKeyPair, chainId: string = this.chainId): string | undefined {
+        assertThrow(!_.isNil(chainId), () => "chain id must be set on signing");
 
-    public withSignature(key: ECKeyPair): Transaction {
         const serializer = new Serializer();
+        const chainBytes = Buffer.from(chainId, "hex");
+
+        const data = serializer.serialize(this);
+        const sig = key.sign(Buffer.concat([chainBytes, data]));
+        return sig;
+    }
+
+    /**
+     * Set a single signature to transaction and return it. May change expiration time to meet valid signature checks for DCore.
+     *
+     * @param key private key
+     * @param chainId id of the DCore chain, different for live/testnet/custom net...
+     */
+    public withSignature(key: ECKeyPair, chainId: string = this.chainId): Transaction {
+        assertThrow(!_.isNil(chainId), () => "chain id must be set on signing");
 
         let sig: string | undefined;
         do {
             // increment expiration until we get dcore valid signature
-            this.blockData.expiration = this.blockData.expiration.add(1, "second");
-            this.expiration = this.blockData.expiration;
-            const data = serializer.serialize(this);
-            sig = key.sign(Buffer.concat([this.chainId, data]));
+            this.expiration = this.expiration.add(1, "second");
+            sig = this.signature(key, chainId);
         } while (!sig);
```

Comment:
What if sig will be undefined infinite times? We should introduce some retry max count and throw error if reached. This applies on all SDKs

Reply:
agree, there should be some `maxExpiration` time for `sign` methods
either you don't care, so we can have some infinite default value,
or you supply some to make sure the expiration does not exceed the threshold

Reply:
Ok